### ISSUE TITLE
[FEATURE REQ] Missing MS_FunctionInvocationId in Error logs

### ISSUE THREAD

**Issue by Stolpe** [Azure/azure-sdk-for-net#32238]
### Library name

Microsoft.Azure.WebJobs.Extensions.ServiceBus

### Please describe the feature.

Hi,

We are missing the MS_FunctionInvocationId property in one of type of error logs thrown by the ServiceBus WebJob extension, namely this one (as described by Serilog):
```
Azure.Messaging.ServiceBus.ServiceBusException: The lock supplied is invalid. Either the lock expired, or the message has already been removed from the queue. For more information please see https://aka.ms/ServiceBusExceptions . Reference:255d0070-bcbf-4e7d-a3c8-b1b29a117075, TrackingId:b0ec5e0e-19cb-473a-ac79-99311818ffcb_B3, SystemTracker:[namespace]:Topic:[topic]|[subscription], Timestamp:2022-11-02T02:26:53 (MessageLockLost)
   at Azure.Messaging.ServiceBus.Amqp.AmqpReceiver.RenewMessageLockInternalAsync(Guid lockToken, TimeSpan timeout)
   at Azure.Messaging.ServiceBus.Amqp.AmqpReceiver.<>c.<<RenewMessageLockAsync>b__59_0>d.MoveNext()
--- End of stack trace from previous location ---
   at Azure.Messaging.ServiceBus.ServiceBusRetryPolicy.RunOperation[T1,TResult](Func`4 operation, T1 t1, TransportConnectionScope scope, CancellationToken cancellationToken, Boolean logRetriesAsVerbose)
   at Azure.Messaging.ServiceBus.ServiceBusRetryPolicy.RunOperation[T1,TResult](Func`4 operation, T1 t1, TransportConnectionScope scope, CancellationToken cancellationToken, Boolean logRetriesAsVerbose)
   at Azure.Messaging.ServiceBus.Amqp.AmqpReceiver.RenewMessageLockAsync(Guid lockToken, CancellationToken cancellationToken)
   at Azure.Messaging.ServiceBus.ServiceBusReceiver.RenewMessageLockAsync(Guid lockToken, CancellationToken cancellationToken)
   at Azure.Messaging.ServiceBus.ServiceBusReceiver.RenewMessageLockAsync(ServiceBusReceivedMessage message, CancellationToken cancellationToken)
   at Azure.Messaging.ServiceBus.ReceiverManager.RenewMessageLockAsync(ServiceBusReceivedMessage message, CancellationTokenSource cancellationTokenSource)
```
The rendering of the log post is as follows:
```
Message processing error (Action=RenewLock, EntityPath=[topic]/Subscriptions/[subscription], Endpoint=[namespace].servicebus.windows.net)
```

Having MS_FunctionInvocationId on these (or any) logs greatly helps troubleshooting.

Have a nice day!

**Comment by jsquire** [comment]
Thank you for your feedback.  Tagging and routing to the team member best able to assist.

**Comment by JoshLove-msft** [comment]
Hi @Stolpe,
The MS_FunctionInvocationId is populated by the Functions framework when a function is executed. This error is not directly tied to a function execution, but rather is tied to lock renewal that the extension is performing in the background.

In terms of tying the function execution and lock renewal together, if you have app insights enabled, you should be able to find a `ServiceBusProcessor.ProcessMessage` span that is the parent of `ServiceBusReceiver.RenewMessageLock`. The full list of instrumented operations can be found [here](https://learn.microsoft.com/en-us/azure/service-bus-messaging/service-bus-end-to-end-tracing?tabs=net-standard-sdk-2#instrumented-operations).

**Comment by Stolpe** [comment]
Hi,

First of all, thank you for your reply. Greatly appreciated!

I understand the technical reasons for why this is. The fact that only code from the Service Bus SDK is in the stack trace is a strong indicator for that. What I had hoped for was that you could somehow rearrange some code so that the logging scope including `MS_FunctionInvocationId` (and surely other properties) would also surround the Service Bus SDK code which performs the RenewLock action (and also _checks logs_ Abandon and ProcessMessageCallback).

Using Application Insights is an interesting workaround. My team and I have set up Application Insights for our WebJobs for a long time, and (when using the latest Nugets) pointed to it using the methods from `Microsoft.Azure.WebJobs.Logging.ApplicationInsights` using just the default configuration and without really knowing how to use it. Maybe this is a good time to learn for real :)

That being said, I tried to locate anything related to your reply, but I was unable to find anything. Again, this could be due to me not really knowing where to look, or could it be that these things are only logged when the code is run as an Azure Function? We run our code as WebJobs in an App Service.

**Comment by JoshLove-msft** [comment]
> That being said, I tried to locate anything related to your reply, but I was unable to find anything. Again, this could be due to me not really knowing where to look, or could it be that these things are only logged when the code is run as an Azure Function? We run our code as WebJobs in an App Service.

Were you able to find any of the [documented operations](https://learn.microsoft.com/en-us/azure/service-bus-messaging/service-bus-end-to-end-tracing?tabs=net-standard-sdk-2#instrumented-operations) in App Insights?

**Comment by Stolpe** [comment]
I was not, but to my defense, there are quite a lot a places to look at :) Some guidance here would be greatly appreciated, but I understand that this might be outside the scope of this issue. Note that this is an Application Insights resource created directly into a Resource Group. The Portal tells me that this is "Classic Application Insights", and that it will be retired in February 2024.
![image](https://user-images.githubusercontent.com/2462597/201337210-8f461e0c-6338-4cd7-bf67-3b77131d1af6.png)

I think I have set up my app correctly. I do see some data that seems to be relevant for my app, but as I said, none of us in my team has ever taken the time to properly learn (and subsequently teach) about what this data actually tells us, and what we can learn from it.
```csharp
Host.CreateDefaultBuilder(args)
    //...
    .ConfigureLogging((context, loggingBuilder) =>
    {
        var configuration = context.Configuration;
        var services = loggingBuilder.Services;

        services.Configure<ApplicationInsightsLoggerOptions>(
            configuration.GetSection(ConfigurationKeys.ApplicationInsights)); //custom location where the Application Insights configuration is stored. only InstrumentationKey is specified for this app

        loggingBuilder
            .ClearProviders() //to remove default console log
            //...
            .AddApplicationInsightsWebJobs();
    })
   //... 
```
By injecting an IOptions<Microsoft.Azure.WebJobs.Logging.ApplicationInsights.ApplicationInsightsLoggerOptions> into the WebJob, I can confirm that the correct InstrumentationKey has indeed been picked up.

**Comment by JoshLove-msft** [comment]
In the [docs](https://learn.microsoft.com/en-us/azure/app-service/webjobs-sdk-get-started#initialize-the-application-insights-logging-provider), they show that you need to pass the instrumentation key into the call to AddApplicationInsightsWebJobs. Can you try that?

**Comment by Stolpe** [comment]
Hi,

Sorry for the late reply. Life got in the way... 🤒👶

I have taken the time to more properly look at Application Insights, and I can confirm that the Application Insights logging does work, even if I set up the `Microsoft.Azure.WebJobs.Logging.ApplicationInsights.ApplicationInsightsLoggerOptions` outside of the `AddApplicationInsightsWebJobs` call. Via the Azure Portal, I entered the Application Insights resource and pressed "Application map". There, I can see a lot of nodes concerning Azure AD, HttpClient, KeyVault, Service Bus, Azure Storage and so on, with my app service in the middle.

I selected the day when these errors last occurred (9 November), ie I use a custom time range from 9 November 12:00 AM to 10 November 12:00 AM. I do not know for sure what time zone this is in, but regardless if it is in my local time (UTC +1) or UTC, my errors should be on the 9th.

I then added a filter where i set "Nodes and Sources, Targets" to equal `ServiceBusReceiver.RenewMessageLock`. This day, we had 3 errors where the Action was `RenewLock`, and 1 each of `Abandon` and `ProcessMessageCallback`. When the filter is applied, the view zooms in to a view where there are just two circles: the app service and an arrow which points to "ServicebusReceiver.RenewMessageLock". The arrow says 49 calls and there is a red exclamation mark above it. I pressed the arrow.

A new view appears to the right, including one button which says "Investigate failures". I pressed that button. Now, if I understand this next view correctly, it says that there was only 1 failure. First of all, I miss two errors, and secondly, I don't know how to get to a view which shows me a parent `ServiceBusProcessor.ProcessMessage`. Under the tab "Dependencies", there is a table of "Top 10 response codes", with only one entry. If I select that entry, and then on the suggested entry again, i get to a view called "End-to-end transaction details". The parent of this node seems to be the app service, but I cannot find `ServiceBusProcessor.ProcessMessage` or anything linking to that.

Do I misunderstand something here, or do I just not understand Application Insights enough? Again, this would be a whole lot simpler for me (and also other people I'd dare say), if we could just get the `FunctionInvocationId` on the log post :)

**Comment by JoshLove-msft** [comment]
> I pressed that button. Now, if I understand this next view correctly, it says that there was only 1 failure. 
It's possible that some transactions are missing if sampling is enabled. 

> Again, this would be a whole lot simpler for me (and also other people I'd dare say), if we could just get the FunctionInvocationId on the log post :)

I'm confused as to what you would do with this information. There isn't actually a function invocation that is tied to the error, as the error does not stem from a function invocation, but rather occurs due to a failure with renewing the lock in the Service Bus library. Do you mean that you want to see any function invocation that might have occurred for the same lock token?  I think we can probably include the lock token in both the exception message from the Service Bus library as well as the Functions trigger logs to help with correlation.

**Comment by JoshLove-msft** [comment]
I just checked and we do already include the lock token for these lock renewal exceptions. Is the log you are referencing in the issue description something you are generating yourself in your error handler?

**Comment by Stolpe** [comment]
Hi,

Sorry for the late reply. I was away on Christmas holidays.

Yes, I have SDK logs enabled. Including the lock token would greatly help with correlation, since that would help me and my team to see what happened during the handling of a particular message.

A little longer version is that we have set up our monitoring to look for logs that have the Error level set. We have a great many systems to monitor, all of which are set up as similarly as possible, in order to reduce complexity. Sometimes we have the case where a single message is handled slowly, then correlation is quite simple. Other times we have tens of thousands messages in queue, and many of them working in parallel. Then it becomes harder to figure out what message failed.

**Comment by Stolpe** [comment]
Hi,

Sorry for the late reply. I was away on Christmas holidays.

Can you clarify what you mean with enabling SDK logging? I add my logger by calling `.ConfigureLogging()` on the `HostBuilder` which later runs the WebJob. Therein I add both my standard logger (Serilog which sinks into an EventHub that dumps logs to an Elasticsearch cluster), and Application Insights from the Microsoft.Azure.WebJobs.Logging.ApplicationInsights nuget.

Does "SDK logging" refer to the standard `ILogger<>` or the Application Insights logger?

**Comment by JoshLove-msft** [comment]
Sorry for the late reply - unless you are manually filtering the SDK logs via host.json, the SDK logs are enabled by default when using the Service Bus functions extension.

**Comment by Stolpe** [comment]
Hi,

My code is run as a WebJob, so I have no host.json file.

I dug up one of these errors from our logs and I have removed some sensitive data from it. As you can see, the error does not include a lock token:
![image](https://user-images.githubusercontent.com/2462597/231397414-a9d70452-a0ec-4918-80eb-a193dbb0f784.png)

**Comment by Stolpe** [comment]
Yes, but this is the only log with the Error log level. No log posts even have the Warning log level, and in my case there are a couple of hundred thousand (Information) log posts in the minutes surrounding the Error log, so finding the function invocation that lies behind this error is really like finding a needle in a haystack. The only realistic way of finding the invocation that lied behind this error is if `MS_FunctionInvocationId` would somehow be added to this log post.

**Comment by JoshLove-msft** [comment]
To clarify, the log I'm referring to is not the same log that you have in the screenshot. The log from the library that includes a lock token is an Error - not Information - https://github.com/Azure/azure-sdk-for-net/blob/main/sdk/servicebus/Azure.Messaging.ServiceBus/src/Diagnostics/ServiceBusEventSource.cs#L688

**Comment by Stolpe** [comment]
Now I feel silly. The reason I did not see that log post was because way back, we decided to filter out log posts from `Azure.Messaging.ServiceBus` (and some additional predicates), since there were so many logs just telling us that it was polling. We just need to modify this filter to only filter out Information log posts, and we will be able to get the lock token, thus enabling us to be able to actually see which invocation caused the problem.

Thank you very much for your time with this. I will go ahead and close this issue now.
